### ISSUE TITLE
Figure out how to handle secrets

### ISSUE THREAD

**Issue by jamesaoverton** [ontodev/droid#45]
Some of the CGI scripts we're developing need access to credentials. Specifically, [COGS](https://github.com/ontodev/cogs) needs to read a JSON file with secrets in order to talk to Google Sheets.

1. My first thought was that DROID can be configured (in `config.edn`) to store some secrets and pass the to Make or CGI scripts in as environment variables. This works fine whether the task is running inside Docker on just on the server. We probably want to be able to do this.

2. Unfortunately the [gpread](https://gspread.readthedocs.io/en/latest/oauth2.html#for-bots-using-service-account) library we're using really expects a file. So my second thought was to create a `projects/FOO/secrets` directory that tasks can just read. But that make it harder for tasks to find the file, and probably means we have to mount another volume for Docker.

3. My current best idea is to create an optional `projects/FOO/configuration/` directory and then copy the contents into the workspace of each branch when the branch is created. (Maybe also on startup, or on file change?) My main worry with this approach is that it's pretty easy to accidentally reveal these files. Maybe DROID could blacklist these files and refuse to serve them. If the scripts/tasks in the repo reveal these credentials, then it's not really DROID's fault.

If it's just COGS that needs this, then maybe should just hack COGS to read an environment variable into the file that it needs.

Better ideas?

**Comment by matentzn** [comment]
I agree it is a pressing issue but no, I don’t have a good idea how to solve this; naively isn’t this something that could be done via a second logging to google, or would this not be sufficient?  I don’t know much about these issues unfortunately.. maybe a private GitHub repo with secrets :P just guessing.

**Comment by jamesaoverton** [comment]
A second log-in to Google is a possibility, but I want to avoid that complexity if possible.

It looks like we can solve the immediate problem using environment variables for COGS, so maybe this just isn't needed. https://github.com/ontodev/cogs/pull/48

**Comment by lmcmicu** [comment]
Proposal 3 seems alright to me.

Regarding James' comment:

> If the scripts/tasks in the repo reveal these credentials, then it's not really DROID's fault

To mitigate this we can make sure those configuration files have standardized names and that those names minimize the chance that someone would inadvertently reveal their contents.

**Comment by jamesaoverton** [comment]
We need this soonish, and it has to work inside Docker containers.

I guess we don't need to have files right now, just environment variables. We have an immediate need for GOOGLE_CREDENTIALS for use with COGS, which is about 10 lines of JSON including a very long line with a private key. This will take an annoying amount of space in config.edn.

Throwing a couple more ideas out there:

4. We could use a secrets management system (e.g. [AWS Secrets Manager](https://aws.amazon.com/secrets-manager/)) configured to the IP of the DROID server.

5. Read/source a `projects/FOO/configuration.env` file. Maybe provide a web-based editing interface for it.

**Comment by jamesaoverton** [comment]
6. [Docker secrets](https://docs.docker.com/engine/swarm/secrets/)

**Comment by lmcmicu** [comment]
Do we really need to hard-code the configuration-file path in DROID? What if, in `config.edn`, we added a configuration parameter (per project) to indicate the location of its configuration file?

In fact we could probably re-use the existing `:env` parameter (if that seems simpler) to handle both actual environment variable definitions and files defining environment variables (see https://github.com/ontodev/droid/blob/8e5d88b71e7677a829d655802c444835586ae548/example-config.edn#L194).

**Comment by jamesaoverton** [comment]
I like the suggestion of `config.edn`'s `:env` (or something nearby) also being able to point to files.

**Comment by jamesaoverton** [comment]
Ok, let's implement reading from a file path into an environment variable and see how far that gets us.

**Comment by jamesaoverton** [comment]
One important case to handle is reading JSON into environment variables. On the main DROID instance, next to the `config.edn` is a `credentials.json` that we've copied into three places in the `config.edn`. When we copied this JSON into the EDN we had to be very careful about escaping double-quotes and double-escaping escaped newlines (`\n`). It would be very nice to just tell DROID to read a JSON file into an environment variable.

**Comment by lmcmicu** [comment]
I have pushed the following commit: https://github.com/ontodev/droid/commit/c5a6be4e85d68d54a586b4e5df0e2a6094f184f4 which will be merged when PR https://github.com/ontodev/droid/pull/108 has been approved. This commit implements the ability to direct DROID to read an environment variable from a file instead of directly.

This works by prefixing the value of the environment variable with "file://"

E.g.,

```
 :projects
 {"lmcmicu-for-testing-with-droid"
  {:project-title "My Project",
   :project-welcome "welcome message",
   :project-description "This project is great",
   :github-coordinates "lmcmicu/for-testing-with-droid",
   :makefile-path nil,
   :env {"REGENV1" "This-is-a-regular-environment-variable"
         "FILENV1" "file://credentials1.json"},
   :docker-config
   {:disabled? false,
    :image "obolibrary/odkfull",
    :workspace-dir "/workspace/",
    :temp-dir "/tmp/droid/",
    :default-working-dir "/workspace/",
    :shell-command "sh",
    :env {"REGENV2" "This-is-another-regular-environment-variable"
          "FILENV2" "file://credentials2.json"}}},
...
```

Let me know if this is sufficient to close this issue.

**Comment by jamesaoverton** [comment]
I like the solution implemented in #108. One shortcoming I'll mention is that setting the `:env` for containers in `config.edn` works fine for shells that DROID runs, but if you connect to the container using `docker exec` then you won't have that environment. That would be nice, but might present new risks. We won't do anything about that right now.
